h5py/_errors.pyx

```diff
@@ -10,76 +10,145 @@
-    H5E_CANTCREATE:     OSError,    # Unable to create file
-    H5E_CANTOPENFILE:   OSError,    # Unable to open file
-    H5E_CANTCLOSEFILE:  OSError,    # Unable to close file
-    H5E_NOTHDF5:        OSError,    # Not an HDF5 file
-    H5E_BADFILE:        ValueError, # Bad file ID accessed
-    H5E_TRUNCATED:      OSError,    # File has been truncated
-    H5E_MOUNT:          OSError,    # File mount error
-
-    H5E_NOFILTER:       OSError,    # Requested filter is not available
-    H5E_CALLBACK:       OSError,    # Callback failed
-    H5E_CANAPPLY:       OSError,    # Error from filter 'can apply' callback
-    H5E_SETLOCAL:       OSError,    # Error from filter 'set local' callback
-    H5E_NOENCODER:      OSError,    # Filter present but encoding disabled
-
-    H5E_BADATOM:        ValueError,  # Unable to find atom information (already closed?)
-    H5E_BADGROUP:       ValueError,  # Unable to find ID group information
-    H5E_BADSELECT:      ValueError,  # Invalid selection (hyperslabs)
-    H5E_UNINITIALIZED:  ValueError,  # Information is uninitialized
-    H5E_UNSUPPORTED:    NotImplementedError,    # Feature is unsupported
-
-    H5E_NOTFOUND:       KeyError,    # Object not found
-    H5E_CANTINSERT:     ValueError,  # Unable to insert object
-
-    H5E_BADTYPE:        TypeError,   # Inappropriate type
-    H5E_BADRANGE:       ValueError,  # Out of range
-    H5E_BADVALUE:       ValueError,  # Bad value
-
-    H5E_EXISTS:         ValueError,  # Object already exists
-    H5E_ALREADYEXISTS:  ValueError,  # Object already exists, part II
-    H5E_CANTCONVERT:    TypeError,   # Can't convert datatypes
-
-    H5E_CANTDELETE:     KeyError,    # Can't delete message
-
-    H5E_CANTOPENOBJ:    KeyError,
-
-    H5E_CANTMOVE:       ValueError,  # Can't move a link
-  }
-
-# "Fudge" table to accommodate annoying inconsistencies in HDF5's use
-# of the minor error codes.  If a (major, minor) entry appears here,
-# it will override any entry in the minor error table.
-_exact_table = {
-    (H5E_CACHE, H5E_BADVALUE):      OSError,  # obj create w/o write intent 1.8
-    (H5E_RESOURCE, H5E_CANTINIT):   OSError,  # obj create w/o write intent 1.6
-    (H5E_INTERNAL, H5E_SYSERRSTR):  OSError,  # e.g. wrong file permissions
-    (H5E_DATATYPE, H5E_CANTINIT):   TypeError,  # No conversion path
-    (H5E_DATASET, H5E_CANTINIT):    ValueError, # bad param for dataset setup
-    (H5E_ARGS, H5E_CANTINIT):       TypeError,  # Illegal operation on object
-    (H5E_SYM, H5E_CANTINIT):        ValueError, # Object already exists/1.8
-    (H5E_ARGS, H5E_BADTYPE):        ValueError, # Invalid location in file
-    (H5E_REFERENCE, H5E_CANTINIT):  ValueError, # Dereferencing invalid ref
-
-    # needed for 1.10.3+ to maintain compatibility with 1.10.{0,1,2}
-
-    # due to changes to H5F.c:H5Fstart_swmr_write
-    (H5E_FILE, H5E_CANTCONVERT):    ValueError, # Invalid file format
-  }
+IF HDF5_VERSION < (1, 13, 0):
```

Comment:
Is it possible to have these `IF` within the long lists, so we don't need to duplicate all the entries that aren't changing? I *think* Cython capitalised `IF` work like C preprocess `#ifdef`, so you can put them in places that wouldn't be valid for a Python `if` statement.

Reply:
(Where there's only a few things in the list, I think it's probably clearer to duplicate them, but for something this long it's hard to see what's different between the `IF` & `ELSE` branches)

Reply:
I wanted to keep the content in sync with `_errors.pxd` and was easier to follow. One block for HDF5-1.13.0 and later, one for before, in both `_errors` files.

Reply:
I'd also prefer having the IF/ELSE only switch what was needed (in both files), and a comment about what changed (e.g. `H5E_ATOM` -> `H5E_ID`). That makes it easy to know what someone needs to search for in the rest of the code (when adding/removing support). The reference for the conditional compilation is https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#compile-time-definitions (I'm not sure through if those definitions are evaluated at the cython -> C transformation time, or if they actually get converted to C preprocessor macros, I think the former, but for h5py currently there's no distinction between the phases).